pkg/networkservice/common/serialize/serializer.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package serialize provides chain elements to make Request, Close processing in chain serial
+package serialize
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/edwarnicke/serialize"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/google/uuid"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+)
+
+type serializer struct {
+	executors map[string]*executor
+	executor  serialize.Executor
+}
+
+type executor struct {
+	executor     serialize.Executor
+	requestCount int
+	// state can be accessed in several ways:
+	// * executor is locked + requestCount is incremented : read, write without atomic
+	// * executor is locked                               : read without atomic, write with atomic
+	// * requestCount == 0                                : read with atomic
+	state uint32
+}
+
+func (s *serializer) requestConnection(
+	ctx context.Context,
+	connID string,
+	requestConn func(context.Context) (*networkservice.Connection, error),
+) (conn *networkservice.Connection, err error) {
+	var requestCh <-chan struct{}
+	var exec *executor
+	<-s.executor.AsyncExec(func() {
+		var ok bool
+		if exec, ok = s.executors[connID]; !ok {
+			exec = new(executor)
+			s.executors[connID] = exec
+		}
+		exec.requestCount++
+
+		requestCh = exec.executor.AsyncExec(func() {
+			for exec.state == 0 {
+				exec.state = uuid.New().ID()
```

Comment:
You are using `uuid` module to generate unique values.

I think the same can be achieved using pointers to empty objects (e.g. `exec.state = &sturct{}{}`). It may be faster and isn't collision prone (as each object would have a pointer that would remain unique until the object is destroyed by the garbage collector).

Reply:
There is a problem with atomic write pointer values, it should be casted to `unsafe.Pointer` and it looks terrible.
Replaced with per-serializer state and atomic increment.